live/liveview.go

```diff
@@ -30,15 +30,112 @@ type Config struct {
 	// PageTitleConfig is a configuration for the page title if the application
 	// uses the liveTitleTag template function in its LayoutTemplate.
 	PageTitleConfig PageTitleConfig
+	// MakeCSRFToken, if non-nil, will be called when a View is requested and should return a valid CSRF token.
+	// If nil, a UUID will be used instead.
+	MakeCSRFToken func(*http.Request) string
 }
 
-func (c *Config) viewForRequest(r *http.Request, currentView View) (View, error) {
-	rw := &JoinHandler{
+type liveViewRequestContextKey struct{}
+type liveViewContainer struct {
+	lv View
+}
+
+func (c *Config) viewForRequest(w http.ResponseWriter, r *http.Request, currentView View) (View, int) {
+	container := &liveViewContainer{
 		lv: currentView,
 	}
+	r = r.WithContext(context.WithValue(r.Context(), liveViewRequestContextKey{}, container))
+
+	rw := &joinHandler{
+		w: w,
+	}
 	c.Mux.ServeHTTP(rw, r)
-	// TODO: check rw.code, convert some status codes (500s?) into meaningful errors
-	return rw.lv, nil
+	return container.lv, rw.code
+}
+
+func (c *Config) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Check for cases where our config's router is its own Mux.
+		// In that case we will re-enter the middleware and should no-op.
+		_, ok := w.(*joinHandler)
+		if ok {
+			next.ServeHTTP(w, r)
+			return
+		}
+		// Get the LiveView if one is routable.
+		lv, code := c.viewForRequest(w, r, nil)
+		if code%100 == 5 {
+			return
+		}
+		if lv == nil {
+			next.ServeHTTP(w, r)
+			return
+		}
+
```

Comment:
I am pretty verbose in comments so feel free to ignore, but I'd probably note something like: "Yay! This route has a LiveView.  Now handle the LiveViewHTTP lifecycle"